models/project.go

```diff
@@ -0,0 +1,300 @@
+	}
+
+	if !IsProjectTypeValid(p.Type) {
+		return errors.New("project type is not valid")
+	}
+
+	sess := x.NewSession()
+	defer sess.Close()
+
+	if err := sess.Begin(); err != nil {
+		return err
+	}
+
+	if _, err := sess.Insert(p); err != nil {
+		return err
+	}
+
+	if _, err := sess.Exec("UPDATE `repository` SET num_projects = num_projects + 1 WHERE id = ?", p.RepoID); err != nil {
+		return err
+	}
+
+	if err := createBoardsForProjectsType(sess, p); err != nil {
+		return err
+	}
+
+	return sess.Commit()
+}
+
+// GetProjectByID returns the projects in a repository
+func GetProjectByID(id int64) (*Project, error) {
+	return getProjectByID(x, id)
+}
+
+func getProjectByID(e Engine, id int64) (*Project, error) {
+	p := new(Project)
+
+	has, err := e.ID(id).Get(p)
+	if err != nil {
+		return nil, err
+	} else if !has {
+		return nil, ErrProjectNotExist{ID: id}
+	}
+
+	return p, nil
+}
+
+// UpdateProject updates project properties
+func UpdateProject(p *Project) error {
+	return updateProject(x, p)
+}
+
+func updateProject(e Engine, p *Project) error {
+	p.UpdatedUnix = timeutil.TimeStampNow()
+	_, err := e.ID(p.ID).Cols(
+		"title",
+		"description",
+		"updated_unix",
+	).Update(p)
+	return err
+}
+
+func countRepoProjects(e Engine, repoID int64) (int64, error) {
+	return e.Where("repo_id=? AND type=?", repoID, ProjectTypeRepository).
+		Count(new(Project))
+}
+
+func countRepoClosedProjects(e Engine, repoID int64) (int64, error) {
+	return e.
+		Where("repo_id=? AND type=? AND is_closed=?", repoID, ProjectTypeRepository, true).
+		Count(new(Project))
+}
+
+// ChangeProjectStatus toggle a project between opened and closed
+func ChangeProjectStatus(p *Project, isClosed bool) error {
+
+	repo, err := GetRepositoryByID(p.RepoID)
+	if err != nil {
+		return err
+	}
+
+	sess := x.NewSession()
+	defer sess.Close()
+	if err = sess.Begin(); err != nil {
+		return err
+	}
+
+	p.IsClosed = isClosed
+	if _, err = sess.ID(p.ID).Cols("is_closed").Update(p); err != nil {
+		return err
+	}
+
+	numProjects, err := countRepoProjects(sess, repo.ID)
+	if err != nil {
+		return err
+	}
+
+	numClosedProjects, err := countRepoClosedProjects(sess, repo.ID)
+	if err != nil {
+		return err
+	}
+
+	repo.NumProjects = int(numProjects)
+	repo.NumClosedProjects = int(numClosedProjects)
+
+	if _, err = sess.ID(repo.ID).Cols("num_projects, num_closed_projects").Update(repo); err != nil {
+		return err
+	}
+
+	return sess.Commit()
+}
+
+// DeleteProjectByID deletes a project from a repository.
+func DeleteProjectByID(id int64) error {
+	sess := x.NewSession()
+	defer sess.Close()
+	if err := sess.Begin(); err != nil {
+		return err
+	}
+
+	if err := deleteProjectByID(sess, id); err != nil {
+		return err
+	}
+
+	return sess.Commit()
+}
+
+func deleteProjectByID(e Engine, id int64) error {
+	p, err := getProjectByID(e, id)
+	if err != nil {
+		if IsErrProjectNotExist(err) {
+			return nil
+		}
+		return err
+	}
+
+	repo, err := getRepositoryByID(e, p.RepoID)
```

Comment:
similar to above - I'm not certain you need to get the repo here.

Reply:
done